google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/test/java/com/google/cloud/spanner/jdbc/ReplaceableForwardingResultSetTest.java

```diff
@@ -0,0 +1,294 @@
+            "getColumnType",
+            "ofResultSet",
+            "equals",
+            "hashCode");
+    ReplaceableForwardingResultSet subject = createSubject();
+    subject.next();
+    subject.close();
+    // Test that all methods throw an SpannerException except the excluded methods when called on a
+    // closed ResultSet.
+    callMethods(subject, excludedMethods, SpannerException.class);
+  }
+
+  @Test
+  public void testMethodCallAfterNextHasReturnedFalse()
+      throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {
+    List<String> excludedMethods =
+        Arrays.asList(
+            "getStats",
+            "next",
+            "close",
+            "getType",
+            "getColumnCount",
+            "getColumnIndex",
+            "getColumnType",
+            "ofResultSet",
+            "equals",
+            "hashCode");
+    ReplaceableForwardingResultSet subject = createSubject();
+    subject.next();
+    subject.next();
+    // Test that all methods throw an IndexOutOfBoundsException except the excluded methods when
+    // called after a call to ResultSet#next() has returned false.
+    callMethods(subject, excludedMethods, IndexOutOfBoundsException.class);
+  }
+
+  private void callMethods(
+      ReplaceableForwardingResultSet subject,
+      List<String> excludedMethods,
+      Class<? extends Exception> expectedException)
+      throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {
+    for (Method method : ReplaceableForwardingResultSet.class.getDeclaredMethods()) {
+      if (Modifier.isPublic(method.getModifiers()) && !excludedMethods.contains(method.getName())) {
+        boolean exception = false;
+        try {
+          if (method.getParameterTypes().length == 0) {
+            method.invoke(subject);
+          } else if (method.getParameterTypes().length == 1
+              && method.getParameterTypes()[0].equals(String.class)) {
+            method.invoke(subject, "test");
+          } else if (method.getParameterTypes().length == 1
+              && method.getParameterTypes()[0].equals(int.class)) {
+            method.invoke(subject, 0);
+          }
+        } catch (InvocationTargetException e) {
+          if (e.getCause().getClass().equals(expectedException)) {
+            // expected
+            exception = true;
+          } else {
+            throw e;
+          }
+        }
+        assertThat(
+            method.getName() + " did not throw an IllegalStateException", exception, is(true));
+      }
+    }
+  }
+
+  @Test
+  public void testValidMethodCall()
+      throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {
+    ResultSet delegate = mock(ResultSet.class);
+    when(delegate.next()).thenReturn(true, true, false);
+    try (ReplaceableForwardingResultSet subject = new ReplaceableForwardingResultSet(delegate)) {
+      subject.next();
+
+      subject.getBoolean(0);
```

Comment:
I'd make a note somewhere in the above implementation that column indexes start at 0

Reply:
I added a note about this in the test case, as I assume that you mean that. In the class itself (`ReplaceableForwardingResultSet`), it is clear from the interface it is implementing that column indices are zero-based.